Tab completion improvement/rewrite

ChaoticMind opened the pull request:
Tab completion can now properly handle elements (most likely usernames) with whitespace. Previously, if someone had a nickname "I like lamp", and one types "I li+tab", it wouldn't work. Another pet peeve tackled!

I'm not super satisfied with it because I would have liked to avoid rebuilding `relevantSubstrings`, but I didn't really find a more elegant solution. (Suggestions welcome). It might be impossible though because of the immutability of strings and the not-worthiness of having a "startsWith()" helper function that works on char arrays.
It's O(n*m) (worst case) where m is the average length of the possible elements (nicknames).

I rewrote a good section of the class to increasing modularity and reduce the need to handle corner cases (although some remnants linger). Commands and Names are now in different `TabCompletionLogic` objects.

ChaoticMind committed:
Rewrite tab completion - now also works on elements with whitespace

ChaoticMind commented:
This class is a nice candidate to write tests on because it has some annoying corner cases that we want to make sure are still working if modifications are made (e.g. one character tab completion, input longer than longest tab completion object, etc). And it's a fairly isolated object with simple string inputs/outputs. I couldn't however figure out how to set up the tests with mono (cmdline linux), any tips?
@RoosterDragon, maybe you can help?

RoosterDragon commented:
I don't think we support that yet sadly: #8849.

Mailaender commented:
Better don't run it on the command line. You are not a CI bot. Use @MonoDevelop: https://github.com/OpenRA/OpenRA/issues/8849#issuecomment-126804827

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs:
```diff
@@ -19,58 +19,83 @@ public class TabCompletionLogic
+			set
+			{
+				universe = value;
+				longestElementLength = universe.Aggregate("", (max, cur) => max.Length > cur.Length ? max : cur).Length;
```
`universe.Max(element => element.Length)` should suffice.

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs:
```diff
@@ -19,58 +19,83 @@ public class TabCompletionLogic
 
-		public IList<string> Commands { get; set; }
+		private int longestElementLength;
+		private IList<string> universe;
```
Our style guide says the default accessibility modifier should be omitted. i.e. You should remove the explicit `private` here.

RoosterDragon commented on OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs:
```diff
@@ -19,58 +19,83 @@ public class TabCompletionLogic
+		}
 
-		public IList<string> Names { get; set; }
+		public TabCompletionLogic(string prefixOnStart = "", string suffixOnStart = ": ")
```
I feel like the empty string would be a better default for the suffix, too.

Phrohdoh replied:
Most clients I've used either append ": " or ", " so one of those two is what I vote for.

RoosterDragon replied:
To be clear, I just mean the default for `TabCompletionLogic` constructor should be an empty string. The caller should explicitly pass the suffix they're after.

atlimit8 commented:
> I couldn't however figure out how to run the current tests with mono (command line on Linux)

See #8861 (reopened)

atlimit8 commented:
> Better don't run it on the command line. You are not a CI bot. Use @MonoDevelop: #8849 (comment)

@Mailaender and what would you put in `git rebase -i --exec 'make ...' bleed` to run the unit tests?

ChaoticMind commented:
Rebased and updated (bugfix)

Covers the following cases, for both starting a string with a nick or not:

1. nicks with and without space
2. nicks with common substring (space or not)
3. nicks with repeating substring (take longest match) (also both with and without space)

Example for the third point, if two nicks are "cabcabal" and "cabal", typing "cabca + tab" will only match the first nick. This is important to cover because it would otherwise get confusing in cases of nicks with spaces.

I had to keep track of "longest match" for each hit, which I did in a Tuple<string, int> (the `string` is the proposed candidate/nick, and the `int` is the length of the longest common substring between the candidate and the input text). I then filtered on the longest match (largest second element).

Another way would be a dictionary of lists, indexed by the "longest common substring length" and then only take the list for the highest index. Unsure if that would be more readable, probably equivalent.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs:
```diff
@@ -20,60 +20,91 @@ public class TabCompletionLogic : ChromeLogic
-		public IList<string> Commands { get; set; }
+		int longestElementLength;
+		IList<string> universe;
+		public IList<string> Universe
```
What is the significance of this name?

ChaoticMind replied:
Set theory [nomenclature](https://en.wikipedia.org/wiki/Universe_%28mathematics%29), seems appropriate.

pchote replied:
This is potentially confusing here, since we are simulating `Actor`s, and `World`s.  Long ago we were going to have a `Universe` class that wrapped the concept of the world and other bits.

atlimit8 replied:
`Pool` or `Completions` should work.

pchote replied:
`Completions` sounds good to me.  It is descriptive and unambigous.

ChaoticMind replied:
Sure thing, I'll wait for the rest of the review and change it.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs:
```diff
@@ -20,60 +20,91 @@ public class TabCompletionLogic : ChromeLogic
 				return text;
 
-			if (lastCompleted == text && candidates.Any())
+			// check if tab is cycled without text modification
```
Start comments with an uppercase letter, please.

ChaoticMind commented:
Updated, rebased, and squashed.

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs:
```diff
@@ -20,60 +20,91 @@ public class TabCompletionLogic : ChromeLogic
+			candidates = new List<string>();
+			var relevantSubstrings =
+				from i in Enumerable.Range(1, Math.Min(text.Length, longestElementLength))
+				select text.Substring(text.Length - i);
```
> [17:40] pchote: functional syntax, please

ChaoticMind replied:
fixed

Mailaender commented:
Works as promised. :white_check_mark:

RoosterDragon commented:
Any chance of some unit tests for this now  #8861 is merged?

atlimit8 commented:
Considering the concern above, it would be nice for the unit tests to be added now rather than later or possibly never. @ChaoticMind use `make nunit` to run NUnit tests in a Linux terminal.

abcdefg30 commented:
Needs a rebase.

obrakmann commented:
Looks good for the most part. I found one issue: when you type "/og", for example, and press tab, it will auto-complete to "/ogiveall", which is obviously wrong.

Phrohdoh commented:
@ChaoticMind Are you still working on this?

ChaoticMind commented:
I haven't worked on this since last commit. It's still on my radar but I can't do it before early February. If you prefer to close it and reopen it when I push an update, that's OK for me. Sorry :(

@obrakmann Yeah, I was planning on making that adjustment in a PR after this one (to keep them smaller). But I guess I may as well just include it in this one.

pchote commented:
Any updates here?

reaperrr commented:
Since there hasn't been any activity for two months, we're closing this to free the PR queue.
@ChaoticMind fell free to reopen it when you get around to update it.
